### PR TITLE
Adding Infested Marines and Infested Bunkers (and a little more)

### PR DESCRIPTION
**Pull request by EnvyDragon**
Generator PR here: https://github.com/Ziktofel/Archipelago/pull/248

New items to unlock:
- Infested Marine, a timed-life unit (90s) buildable from the Infested Barracks.
- Infested Bunker, comes with 4 Infested Troopers, can uproot to move around and attack. While rooted, regenerates health and spawns new troopers every 30s.
- Sustained Cultivation Ventricles (Infested SCV): Increases InfSCV build charges to 3 starting, 5 max.
- Plagued Munitions (Infested infantry): Adds a damage over time effect to the weapons of Infested Marines, Infested Troopers, and Infestor Infested Terrans.
- Retinal Augmentation (Infested infantry): Adds +1 range to the weapons of Infested Marines/Troopers/Infestor Terrans.
- Calcified Armor (Infested Bunker): Adds +3 armor to the Infested Bunker.
- Regenerative Plating (Infested Bunker): Increases Infested Bunker regen rate while rooted.
- Engorged Bunkers (Infested Bunker): +2 cargo space and 20% reduced cooldown on trooper spawning for Infested Bunkers.

Supporting units (ie. not items to be found in the world, but either unlocked or produced by the above):
- Infested SCV, buildable from Hatchery/Lair/Hive on a charge system. Starts at 1 charge to limit arbitrary boost to economic power, with an upgrade that can increase it to the Stukov Co-op levels (3 starting charges, 5 max).
- Infested Trooper, a timed-life unit (30s) spawned automatically from the Infested Bunker. Almost entirely identical to the Infested Marine.
- Infested Barracks, stores 10 charges of Infested Marines. Can uproot/root to move around.
- Infested Cocoons, which InfSCVs and Infested Marines spawn from. 

Triggers:
- When Infested Bunkers are built, they are filled with Infested Troopers up to their cargo capacity.
- When Infested Bunkers fire their periodic "spawn" dummy behavior, a new Trooper is either added to their cargo or spawned outside, moving to a rally point if applicable.
- When either of the above triggers occur, the spawned troopers get a timed life behavior applied to them.

### CODE REVIEW

**Comment by Ziktofel** [comment]
Snímek obrazovky pořízený 2024-07-24 21-29-50
Snímek obrazovky pořízený 2024-07-24 21-29-32
Bug: Infested SCV gets the attack speed upgrade from normal SCV (Hostile Environment Adaptation) but not its HP or displayed icon

**Comment by Ziktofel** [comment]
Normal SCVs autocast repairs. Infested don't
Snímek obrazovky pořízený 2024-07-24 21-32-11

**Comment by Ziktofel** [comment]
Script Error:
Snímek obrazovky pořízený 2024-07-24 21-35-14

**Comment by Ziktofel** [comment]
Snímek obrazovky pořízený 2024-07-24 21-46-02
Missing Actor `AP_SIInfestedMarine_Land`

**Comment by Ziktofel** [comment]
Snímek obrazovky pořízený 2024-07-24 21-47-51
Wild Mutation (Shield name/color needs to be copied from other Zerg units like lings)

**Comment by Ziktofel** [comment]
Snímek obrazovky pořízený 2024-07-24 22-02-29
Snímek obrazovky pořízený 2024-07-24 22-02-18
Add a requirement for Rooted for display (the same way as Tank use). Also do this for Regenerative Plating icon

**Review by Ziktofel — CHANGES_REQUESTED** [review]
Fix the bugs from above

**Comment by EnvyDragon** [comment]
> Script Error: Snímek obrazovky pořízený 2024-07-24 21-35-14 Triggers when Infested Bunker completes `Spawn Infested Trooper` and rally point is not set

This error is harmless and is accounted for in the triggers. Unfortunately the sc2 editor doesn't appear to have a try/catch method, so if I use the only function in existence that can get the target point of a rally ability, it throws an error if there's no rally point. I'll take a second look, but the script error doesn't show up in regular gameplay and the trigger is designed to accommodate a missing point even if the editor finds it upsetting.

**Comment by EnvyDragon** [comment]
> Add a requirement for Rooted for display (the same way as Tank use). Also do this for Regenerative Plating icon

What you're seeing is the vanilla behavior. The icon is present-but-disabled when uprooted so it's clearer to players that it won't spawn if it's mobile. I can put the same visual effect on Regenerative Plating though, if it helps.

**Comment by Ziktofel** [comment]
> > Script Error: Snímek obrazovky pořízený 2024-07-24 21-35-14 Triggers when Infested Bunker completes `Spawn Infested Trooper` and rally point is not set
> 
> This error is harmless and is accounted for in the triggers. Unfortunately the sc2 editor doesn't appear to have a try/catch method, so if I use the only function in existence that can get the target point of a rally ability, it throws an error if there's no rally point. I'll take a second look, but the script error doesn't show up in regular gameplay and the trigger is designed to accommodate a missing point even if the editor finds it upsetting.

Add a check if the point exists and if not, use Skip Remaining Actions

**Comment by Ziktofel** [comment]
> > Add a requirement for Rooted for display (the same way as Tank use). Also do this for Regenerative Plating icon
> 
> What you're seeing is the vanilla behavior. The icon is present-but-disabled when uprooted so it's clearer to players that it won't spawn if it's mobile. I can put the same visual effect on Regenerative Plating though, if it helps.

Use the same visual effect + like tanks do, use the requirement that tells the player that the infested bunker needs to be rooted. It's used for Tanks/Liberators/Vikings where a trait works only in one mode

**Comment by EnvyDragon** [comment]
> > > Add a requirement for Rooted for display (the same way as Tank use). Also do this for Regenerative Plating icon
> > 
> > 
> > What you're seeing is the vanilla behavior. The icon is present-but-disabled when uprooted so it's clearer to players that it won't spawn if it's mobile. I can put the same visual effect on Regenerative Plating though, if it helps.
> 
> Use the same visual effect + like tanks do, use the requirement that tells the player that the infested bunker needs to be rooted. It's used for Tanks/Liberators/Vikings where a trait works only in one mode

Not sure what you mean or what you're suggesting, so I just did this:
image

**Comment by EnvyDragon** [comment]
> > > Script Error: Snímek obrazovky pořízený 2024-07-24 21-35-14 Triggers when Infested Bunker completes `Spawn Infested Trooper` and rally point is not set
> > 
> > 
> > This error is harmless and is accounted for in the triggers. Unfortunately the sc2 editor doesn't appear to have a try/catch method, so if I use the only function in existence that can get the target point of a rally ability, it throws an error if there's no rally point. I'll take a second look, but the script error doesn't show up in regular gameplay and the trigger is designed to accommodate a missing point even if the editor finds it upsetting.
> 
> Add a check if the point exists and if not, use Skip Remaining Actions

Skip Remaining Actions isn't an option because it needs to do a bunch of other actions, like spawning the actual unit. I found a workaround though, the error warning is now gone.

**Comment by Ziktofel** [comment]
> > > > Add a requirement for Rooted for display (the same way as Tank use). Also do this for Regenerative Plating icon
> > > 
> > > 
> > > What you're seeing is the vanilla behavior. The icon is present-but-disabled when uprooted so it's clearer to players that it won't spawn if it's mobile. I can put the same visual effect on Regenerative Plating though, if it helps.
> > 
> > 
> > Use the same visual effect + like tanks do, use the requirement that tells the player that the infested bunker needs to be rooted. It's used for Tanks/Liberators/Vikings where a trait works only in one mode
> 
> Not sure what you mean or what you're suggesting, so I just did this: image

Snímek obrazovky pořízený 2024-07-25 18-52-35
Add this (it's a requirement that leads to an upgrade that's never researched). All transformers in AP use this

**Comment by Ziktofel** [comment]
[
Snímek obrazovky pořízený 2024-07-25 18-57-00
Snímek obrazovky pořízený 2024-07-25 18-56-02
](url)
Infested SCVs are affected by Terran Advanced Construction upgrade but the player isn't notified about that

File: Mods/ArchipelagoPlayer.SC2Mod/Base.SC2Data/GameData/UnitData.xml [hunk 1]
```diff
@@ -21442,6 +21462,413 @@
...
+        <AbilArray Link="AP_SCVHarvest"/>
+        <AbilArray Link="AP_SIAdvancedBuild"/>
+        <WeaponArray Link="AP_SISCVFusionCutter"/>
+        <CardLayouts>
+            <LayoutButtons Face="Move" Type="AbilCmd" AbilCmd="move,Move" Row="0" Column="0"/>
+            <LayoutButtons Face="Stop" Type="AbilCmd" AbilCmd="stop,Stop" Row="0" Column="1"/>
+            <LayoutButtons Face="MoveHoldPosition" Type="AbilCmd" AbilCmd="move,HoldPos" Row="0" Column="2"/>
+            <LayoutButtons Face="AttackWorker" Type="AbilCmd" AbilCmd="attack,Execute" Row="0" Column="4"/>
+            <LayoutButtons Face="MovePatrol" Type="AbilCmd" AbilCmd="move,Patrol" Row="0" Column="3"/>
+            <LayoutButtons Face="Repair" Type="AbilCmd" AbilCmd="Repair,Execute" Row="2" Column="2"/>
```
Comment (Ziktofel) [inline Mods/ArchipelagoPlayer.SC2Mod/Base.SC2Data/GameData/UnitData.xml:21706]: AP_Repair

File: Mods/ArchipelagoPlayer.SC2Mod/Base.SC2Data/GameData/UnitData.xml [hunk 2]
```diff
@@ -21442,6 +21462,413 @@
...
+            <LayoutButtons Face="GatherTerr" Type="AbilCmd" AbilCmd="SCVHarvest,Gather" Row="1" Column="0"/>
+            <LayoutButtons Face="ReturnCargo" Type="AbilCmd" AbilCmd="SCVHarvest,Return" Row="1" Column="1"/>
```
Comment (Ziktofel) [inline Mods/ArchipelagoPlayer.SC2Mod/Base.SC2Data/GameData/UnitData.xml:21708]: AP_SCVHarvest

**Review by Ziktofel — CHANGES_REQUESTED** [review]

**Comment by EnvyDragon** [comment]
> Snímek obrazovky pořízený 2024-07-25 18-52-35 Add this (it's a requirement that leads to an upgrade that's never researched). All transformers in AP use this

image
Added! Will add the SCV upgrades then commit.

**Comment by Ziktofel** [comment]
Also if your Infested Barracks is uprooted, you can no longer build infested Bunkers (normally lifted off buildings do count for tech requirements)
Snímek obrazovky pořízený 2024-07-25 19-17-01

**Comment by EnvyDragon** [comment]
image
Applied SCV upgrades to infested SCVs

**Comment by EnvyDragon** [comment]
Uprooted Barracks now uses the same tech alias as the rooted version (pictured is the only infested barracks on the map)
image

**Pull request merged**